### PR TITLE
[FMV] Change the default versions name mangling.

### PR DESCRIPTION
**Pull request by DanielKristofKiss**
The default version's name to be mangled to avoid problems with ifuncs due to the ifunc to be called as the base name of the function.

As for any pull request, please make sure to go through the below
checklist.

Checklist: (mark with ``X`` those which apply)

* [ ] If an issue reporting the bug exists, I have mentioned it in the
      PR (do not bother creating the issue if all you want to do is
      fixing the bug yourself).
* [ ] I have added/updated the `SPDX-FileCopyrightText` lines on top
      of any file I have edited. Format is `SPDX-FileCopyrightText:
      Copyright {year} {entity or name} <{contact informations}>`
      (Please update existing copyright lines if applicable. You can
      specify year ranges with hyphen , as in `2017-2019`, and use
      commas to separate gaps, as in `2018-2020, 2022`).
* [ ] I have updated the `Copyright` section of the sources of the
      specification I have edited (this will show up in the text
      rendered in the PDF and other output format supported). The
      format is the same described in the previous item.
* [ ] I have run the CI scripts (if applicable, as they might be
      tricky to set up on non-*nix machines). The sequence can be
      found in the [contribution
      guidelines](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#continuous-integration). Don't
      worry if you cannot run these scripts on your machine, your
      patch will be automatically checked in the Actions of the pull
      request.
* [x] I have added an item that describes the changes I have
      introduced in this PR in the section **Changes for next
      release** of the section **Change Control**/**Document history**
      of the document. Create **Changes for next release** if it does
      not exist. Notice that changes that are not modifying the
      content and rendering of the specifications (both HTML and PDF)
      do not need to be listed.
* [x] When modifying content and/or its rendering, I have checked the
      correctness of the result in the PDF output (please refer to the
      instructions on [how to build the PDFs
      locally](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#continuous-integration)).
* [x] The variable `draftversion` is set to `true` in the YAML header
      of the sources of the specifications I have modified.
* [ ] Please *DO NOT* add my GitHub profile to the list of contributors
      in the [README](https://github.com/ARM-software/acle/blob/main/README.md#contributors-) page of the project.

### CODE REVIEW

**Comment by vhscampos** [comment]
Some points aren't clear in the text:
1. Will this new `default` name mangling happen only with FMV enabled or globally?
2. If the _only_ version specified is the `default` one, either explicitly with the attribute or implictly without the attribute, will the new mangling still take place?

**Comment by DanielKristofKiss** [comment]
> Some points aren't clear in the text:
> 
> 1. Will this new `default` name mangling happen only with FMV enabled or globally?

Only when FMV enabled. By amending the `default` version with `"default"` the basename could be used by the resolved symbol, referenced externally as it won't collide with the default function.

> 2. If the _only_ version specified is the `default` one, either explicitly with the attribute or implictly without the attribute, will the new mangling still take place?

Depends on, for the same as above reasons. if the toolchain can avoid emitting redirections then no need to use the default version. maybe a alias could be useful.

File: main/acle.md [hunk 1]
```diff
@@ -2474,8 +2476,9 @@ compiler and it is enabled.
 
 ### Name mangling
 
-The `"default"` version is not mangled on top of the language-specific name
-mangling.
+The `"default"` version is mangled with `".default"` on top of the
+language-specific name mangling. When the `"default"` version matches with an
+explicitly provided version an alias to be generated.
```
Comment (DanielKristofKiss) [inline main/acle.md:2481]:
```suggestion
explicitly provided version an alias to be generated.
All versions expected to have at least alias that points to actual version.
```

**Comment by andrewcarlotti** [comment]
I don't understand the remark about aliases in this change. Is this saying that the compiler can optimise versions with identical function bodies to a single function with additional aliased names? If so, I don't think that's relevant to this spec. Alternatively, is it referring to a case where a non-default version specifies only features that the compiler is already assuming to be present?

**Comment by andrewcarlotti** [comment]
Aside from that, I think the mangling changes are reasonable. It might be worth adding an explicit remark that the versioned symbol (i.e. the one that is resolved to one of the versions) should have the same mangling as would be used if the function didn't use function multiversioning.

**Comment by DanielKristofKiss** [comment]
> I don't understand the remark about aliases in this change. Is this saying that the compiler can optimise versions with identical function bodies to a single function with additional aliased names? If so, I don't think that's relevant to this spec. Alternatively, is it referring to a case where a non-default version specifies only features that the compiler is already assuming to be present?

The intention is to be sure all function names remain always available when the certain optimisation eliminates a version. (e.g. the default and a given version got merged).

> Aside from that, I think the mangling changes are reasonable. It might be worth adding an explicit remark that the versioned symbol (i.e. the one that is resolved to one of the versions) should have the same mangling as would be used if the function didn't use function multiversioning.

ACK.

File: main/acle.md [hunk 2]
```diff
@@ -2496,8 +2499,9 @@ For example:
 __attribute__((target_clones("crc32", "aes+sha1")))
 int foo(){..}
 ```
-will produce these mangled names for C language: `foo`, `foo._Mcrc32`,
-`foo._Msha1Maes`.
+will produce these mangled names for C language: `foo.default`, `foo._Mcrc32`,
+`foo._Msha1Maes` while `foo` is a callable external symbol which leads to one
+of the versioned function.
```
Comment (vhscampos) [inline main/acle.md:2504]: Nit: "versioned function" -> "version functions". Plural

File: main/acle.md [hunk 3]
```diff
@@ -2476,8 +2477,10 @@ compiler and it is enabled.
 
 ### Name mangling
 
-The `"default"` version is not mangled on top of the language-specific name
-mangling.
+The `"default"` version is mangled with `".default"` on top of the
+language-specific name mangling.
+All versioned functions with their mangled name are always resolvable.
```
Comment (vhscampos) [inline main/acle.md:2482]: Nit: "name" -> "names". Plural

Comment (vhscampos) [inline main/acle.md:2482]: I suggest that you move this line to the end of the previous one, i.e. make the two form a single paragraph.

**Review by vhscampos — CHANGES_REQUESTED** [review]

**Comment by andrewcarlotti** [comment]
This looks good to me now. I've already implemented this mangling in GCC.

**Review by vhscampos — APPROVED** [review]

**Review by sallyarmneale — COMMENTED** [review]
Minor changes

**Pull request merged**